### PR TITLE
New Resource: `azurerm_linux_virtual_machine_scale_set`

### LINKED ISSUE #160: VMSS: support for specifying the OS Disk Size
### Terraform Version
Terraform v0.9.11

### Affected Resource(s)
- azurerm_virtual_machine_scale_set

There are several things that aren't available for azurerm_virtual_machine_scale_set but are available for individual VMs.  I have done all of them using ARM templates so I know it's possible in Azure, but terraform doesn't allow them (I tried :)

Please let me know if I should split this out into separate requests for each.

1. ~Would it be possible to support a boot_diagnostics block in the VMSS configuration, just like the one in azurerm_virtual_machine?  See the "How do I turn on boot diagnostics" section here: https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-faq~ (@tombuildsstuff: this is now available)

2. ~Could you please add support for network security groups in a VMSS?  This is already there for azurerm_network_interface but not for the scale set.  It seems like it would just be a network_security_group_id field inside the VMSS network_profile.~ (@tombuildsstuff: this is now available)

3. Could you please add the ability to specify a size for OS Disks?  In azurerm_virtual_machine there is a storage_os_disk block where you can specify disk_size_gb, but with the VMSS the storage_profile_**os**_disk block does not support the disk_size_gb field (the VMSS storage_profile_data_disk does support disk_size_gb but I need the OS disk).

Thank you!

### LINKED ISSUE #292: Issue with nested attributes
### Terraform Version
Terraform v0.10.2

### Affected Resource(s)
- azurerm_virtual_machine_scale_set
- azurerm_container_service

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
We used this as our starting point: https://www.terraform.io/docs/providers/azurerm/r/virtual_machine_scale_set.html
```hcl
# Started with this:
network_profile {
    name    = "TestNetworkProfile"
    primary = true

    ip_configuration {
      name      = "TestIPConfiguration"
      subnet_id = "${azurerm_subnet.test.id}"
    }
  }

# Then tried to update to this:
network_profile {
    name    = "terraformnetworkprofile"
    primary = true

    ip_configuration {
      name                                   = "TestIPConfiguration"
      subnet_id                              = "${azurerm_subnet.test.id}"
      load_balancer_backend_address_pool_ids = ["${azurerm_lb_backend_address_pool.bpepool.id}"]
      load_balancer_inbound_nat_rules_ids    = ["${element(azurerm_lb_nat_pool.lbnatpool.*.id, count.index)}"]
    }
  }
```

### Panic Output
N/A

### Expected Behavior
For all attributes of resources, changing the value should result in TF plan showing a required change to the resource, and TF apply should update the resource.

### Actual Behavior
Running TF plan after changing any of the following "nested" attributes shows 'No changes. Infrastructure is up-to-date.' Running TF apply makes no changes to these existing resources.
- azurerm_virtual_machine_scale_set/os_profile_linux_config/ssh_keys/key_data
- azurerm_virtual_machine_scale_set/network_profile/ip_configuration/load_balancer_inbound_nat_rules_ids
- azurerm_container_service/linux_profile/ssh_key/key_data

### Steps to Reproduce
1. `terraform apply`
2. Change the value of os_profile_linux_config/ssh_keys/key_data
3. `terraform plan`
4. `terraform apply`
5. Notice that key_data did not change. 

### References
N/A

### LINKED ISSUE #536: Feature req: Allow for 'soft' image updates to VMSS
Presently, image updates to a VMSS trigger a full teardown & rebuild of the VMSS (and its instances) to match the configuration.

There are valid use cases where a non-destructive image update is preferred. In this instance the VMSS is updated with the appropriate image, but underlying VMSS instances are not automatically torn down and rebuilt as they may be in-use. Often auto-updating VMSS instances is left up to systems external to TF as they safely upgrade instances when appropriate.

I would propose some additional flag to allow this behavior to be 'soft' as an override, where the default behavior would be as it is today.

### LINKED ISSUE #809: azurerm_virtual_machine_scale_set.scaleset: diffs didn't match during apply
_This issue was originally opened by @agolomoodysaada as hashicorp/terraform#17291. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

```
Terraform Version: 0.11.3
    Resource ID: azurerm_virtual_machine_scale_set.scaleset
    Mismatch reason: attribute mismatch: network_profile.3458559321.ip_configuration.#
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"network_profile.~3458559321.ip_configuration.0.primary":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.accelerated_networking":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.ip_configuration.0.load_balancer_backend_address_pool_ids.#":*terraform.ResourceAttrDiff{Old:"0", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.ip_configuration.0.public_ip_address_configuration.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.ip_configuration.0.load_balancer_backend_address_pool_ids.2187753327":*terraform.ResourceAttrDiff{Old:"/subscriptions/93c2ebb5-31e9-487d-9f19-f1716b0673ce/resourceGroups/historical-filters/providers/Microsoft.Network/loadBalancers/agolo-staging-historical-filters/backendAddressPools/BackEndAddressPool", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.ip_configuration.0.load_balancer_backend_address_pool_ids.#":*terraform.ResourceAttrDiff{Old:"1", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.primary":*terraform.ResourceAttrDiff{Old:"true", New:"false", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.ip_configuration.0.name":*terraform.ResourceAttrDiff{Old:"", New:"historical-filters-ipconfig", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.primary":*terraform.ResourceAttrDiff{Old:"", New:"true", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.ip_configuration.#":*terraform.ResourceAttrDiff{Old:"1", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.network_security_group_id":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.accelerated_networking":*terraform.ResourceAttrDiff{Old:"false", New:"false", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.ip_configuration.0.subnet_id":*terraform.ResourceAttrDiff{Old:"", New:"/subscriptions/93c2ebb5-31e9-487d-9f19-f1716b0673ce/resourceGroups/general/providers/Microsoft.Network/virtualNetworks/agolo/subnets/historical-filters-subnet", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.ip_configuration.0.primary":*terraform.ResourceAttrDiff{Old:"false", New:"false", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.ip_configuration.#":*terraform.ResourceAttrDiff{Old:"0", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil),RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.ip_configuration.0.public_ip_address_configuration.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.ip_configuration.0.subnet_id":*terraform.ResourceAttrDiff{Old:"/subscriptions/93c2ebb5-31e9-487d-9f19-f1716b0673ce/resourceGroups/general/providers/Microsoft.Network/virtualNetworks/agolo/subnets/historical-filters-subnet", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.name":*terraform.ResourceAttrDiff{Old:"", New:"historical-filters-nic", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.network_security_group_id":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.ip_configuration.0.load_balancer_backend_address_pool_ids.2187753327":*terraform.ResourceAttrDiff{Old:"", New:"/subscriptions/93c2ebb5-31e9-487d-9f19-f1716b0673ce/resourceGroups/historical-filters/providers/Microsoft.Network/loadBalancers/agolo-staging-historical-filters/backendAddressPools/BackEndAddressPool", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.~3458559321.ip_configuration.0.load_balancer_inbound_nat_rules_ids.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.ip_configuration.0.name":*terraform.ResourceAttrDiff{Old:"historical-filters-ipconfig", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "network_profile.3458559321.name":*terraform.ResourceAttrDiff{Old:"historical-filters-nic", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff(nil), Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}
```

### LINKED ISSUE #1616: azurerm_virtual_machine_scale_set: mismatch on apply
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.7
+ provider.azurerm v1.9.0
+ provider.random v1.3.1
+ provider.template v1.0.0

### Affected Resource(s)

* azurerm_virtual_machine_scale_set

### Terraform Configuration Files

Output of terraform plan and apply attached:
[terraform.bugreport.txt.asc.txt](https://github.com/terraform-providers/terraform-provider-azurerm/files/2211326/terraform.bugreport.txt.asc.txt)

### Expected Behavior

Application of plan without error

### Actual Behavior

    Terraform Version: 0.11.7
    Resource ID: azurerm_virtual_machine_scale_set.app
    Mismatch reason: attribute mismatch: network_profile.1854554140.ip_configuration.0.application_gateway_backend_address_pool_ids.1842515611

### Steps to Reproduce

1. `terraform apply`

### References

* Ø

### LINKED ISSUE #1620: tried to add load balancer backend pools to existing VMs, terraform spat out a "I have a bug" report
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.7
+ provider.azurerm v1.9.0
+ provider.random v1.3.1
+ provider.template v1.0.0

### Affected Resource(s)

* azurerm_virtual_machine_scale_set

### Terraform Configuration Files

(available on request)

### Debug Output

Full terraform output attached:

[terraform-bugreport-mismatch.txt.asc.txt](https://github.com/terraform-providers/terraform-provider-azurerm/files/2215287/terraform-bugreport-mismatch.txt.asc.txt)

### Expected Behavior

`terraform apply` should have made the necessary changes

### Actual Behavior

Changes seem to not be applied, also:

```
* module.meta_cluster.azurerm_virtual_machine_scale_set.app: azurerm_virtual_machine_scale_set.app: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.
```

### Steps to Reproduce

1. `terraform apply`

### References

* Ø

### LINKED ISSUE #1669: azurerm_virtual_machine_scale_set does not properly remove data disks
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.7
+ provider.azurerm v1.11.0
+ provider.random v1.3.1
+ provider.template v1.0.0

### Affected Resource(s)

* azurerm_virtual_machine_scale_set

### Terraform Configuration Files

old:

```hcl
azurerm_virtual_machine_scale_set {
  …
  # data disks
  storage_profile_data_disk {
    managed_disk_type = "Standard_LRS"
    create_option     = "Empty"
    lun               = 0
    disk_size_gb      = "32"
  }
}
```

new: (same as old but with no `storage_profile_data_disk` block)

### Expected Behavior

terraform should update VMSS configuration to remove the data disk

### Actual Behavior

```text
  ~ module.meta_cluster.azurerm_virtual_machine_scale_set.app
      storage_profile_data_disk.#:               "1" => "0"
      storage_profile_data_disk.0.create_option: "Empty" => ""
```

result:
```
module.meta_cluster.azurerm_autoscale_setting.app: Modifications complete after 5s (ID: /subscriptions/16b9f5dc-3276-4b32-91b2-.../autoscalesettings/metaDefaultSettings)
```

new plan:
```
  ~ module.meta_cluster.azurerm_virtual_machine_scale_set.app
      storage_profile_data_disk.#:               "1" => "0"
      storage_profile_data_disk.0.create_option: "Empty" => ""
```

The actual storage profile hasn't changed.

### Steps to Reproduce

1. `terraform apply` the config with a data disk
1. `terraform apply` the config without a data disk

### References

* Ø

### LINKED ISSUE #1739: "diffs didn't match during apply" when updating Scale Set
### Community Note

* Please vote on this issue by adding a :+1: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.7
+ provider.azurerm v1.12.0
```

### Affected Resource(s)

* azurerm_virtual_machine_scale_set

### Terraform Configuration Files

```hcl
resource "azurerm_resource_group" "sample" {
  name = "sample"
  location = "Japan East"
}

resource "azurerm_virtual_network" "sample" {
  name = "sample"
  resource_group_name = "${azurerm_resource_group.sample.name}"
  location = "${azurerm_resource_group.sample.location}"
  address_space = ["10.0.0.0/16"]
}

resource "azurerm_subnet" "sample" {
  name = "sample"
  resource_group_name = "${azurerm_resource_group.sample.name}"
  virtual_network_name = "${azurerm_virtual_network.sample.name}"
  address_prefix = "10.0.2.0/24"
}

resource "azurerm_virtual_machine_scale_set" "sample" {
  name = "sample"
  resource_group_name = "${azurerm_resource_group.sample.name}"
  location = "${azurerm_resource_group.sample.location}"
  upgrade_policy_mode = "Manual"

  sku {
    name = "Standard_DS1_v2"
    tier = "Standard"
    capacity = 1
  }

  storage_profile_image_reference {
    publisher = "Canonical"
    offer = "UbuntuServer"
    sku = "16.04.0-LTS"
    version = "latest"
  }

  storage_profile_os_disk {
    name = ""
    caching = "ReadWrite"
    create_option = "FromImage"
    managed_disk_type = "Premium_LRS"
  }

  storage_profile_data_disk {
    lun = 0
    caching = "ReadWrite"
    create_option = "Empty"
    disk_size_gb = 10
  }

  os_profile {
    computer_name_prefix = "sample"
    admin_username = "ubuntu"
    admin_password = "password"
  }

  os_profile_linux_config {
    disable_password_authentication = true
    ssh_keys {
      path = "/home/ubuntu/.ssh/authorized_keys"
      key_data = "${file("~/.ssh/azure_default.pub")}"
    }
  }

  network_profile {
    name = "TestNetworkProfile"
    primary = true
    ip_configuration {
      name = "TestIPConfiguration"
      subnet_id = "${azurerm_subnet.sample.id}"
    }
    # network_security_group_id = "${azurerm_network_security_group.sample.id}"
  }
}

# resource "azurerm_network_security_group" "sample" {
#   name = "sample"
#   resource_group_name = "${azurerm_resource_group.sample.name}"
#   location = "${azurerm_resource_group.sample.location}"

#   security_rule {
#     name = "SSH"
#     priority = 1001
#     direction = "Inbound"
#     access = "Allow"
#     protocol = "Tcp"
#     source_port_range = "*"
#     destination_port_range = "22"
#     source_address_prefix = "*"
#     destination_address_prefix = "*"
#   }
# }
```

### Debug Output
https://gist.github.com/shibataka000/872f7336be98a8e4be277fa6933b9779

### Panic Output

n/a

### Expected Behavior

Terraform should have updated the scale set with the network security group.

### Actual Behavior

Got the "diffs didn't match during apply" error shown in the gist.

### Steps to Reproduce

1. `terraform apply`.
2. Revert following statement which is commented out.
    - `resource "azurerm_network_security_group" "sample" { ... }`
    - `network_security_group_id = "${azurerm_network_security_group.sample.id}"`
3. `terraform apply` again.
4. Get the above error.

### LINKED ISSUE #3020: azurerm_virtual_machine_scale_set support for public ip per instance parameter
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
Terraform doesn't appear to support the use of configuring the Azure Scale Set with a parameter to enable a public IP address per instance. This is available to configure via the Azure GUI / CLI.

### Affected Resources
azurerm_virtual_machine_scale_set

### Versions
Terraform: v0.11.11
Azurerm: v0.11.0

### LINKED ISSUE #3773: Add support for VMSS Ephemeral OS Disk
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add support for putting the OS disk on the ephemeral drive

### New or Affected Resource(s)

* azurerm_virtual_machine_scale_set

### Potential Terraform Configuration

```hcl
resource "azurerm_virtual_machine_scale_set" "test_vmss" {
  # all the normal stuff plus storage_profile_os_disk -> diff_disk_settings
  storage_profile_os_disk {
    caching            = "ReadOnly"
    create_option      = "FromImage"
    diff_disk_settings {
      option = "Local"
    }
  }
}
```

### References

* https://azure.microsoft.com/en-us/blog/ephemeral-os-disk-limited-public-preview/
* https://docs.microsoft.com/en-us/azure/templates/microsoft.compute/2019-03-01/virtualmachinescalesets
* https://github.com/Azure/azure-cli/issues/7719

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR adds support for one of the 2.0 resources: `azurerm_linux_virtual_machine_scale_set` which supersedes [the existing `azurerm_virtual_machine_scale_set` resource](https://www.terraform.io/docs/providers/azurerm/r/virtual_machine_scale_set.html) as outlined in #2807.

The original `azurerm_virtual_machine_scale_set` resource was introduced in March 2016, when Azure Resource Manager was still in Beta.

Whilst we've been adding features since then, unfortunately the way the Virtual Machine Scale Set API behaves today is very different from how it did previously - as such the `azurerm_virtual_machine_scale_set` resource didn't fully cover all use-cases.

After spending a considerable amount of time on this - we ultimately came to the conclusion it'd be better to start afresh rather than trying to fix the existing `azurerm_virtual_machine_scale_set` resource. The Azure API behaves differently depending on which Operating System is used on the Virtual Machine / Virtual Machine Scale Set - for example for a Windows Virtual Machine (& Scale Set) a `name` field can be up to 15 Characters long, whereas for a Linux Virtual Machine (& Scale Set) the `name` field can be up to 63 characters long.

---

There's some intentional design decisions as a part of this resource:

* Unmanaged Disks are not supported by this new resource since Azure no longer recommends Unmanaged Disks for new deployments (and there's considerable scaling limitations when using them). Operators wishing to use Unmanaged Disks can continue to use the existing `azurerm_virtual_machine_scale_set` resource.
* Terraform will now update and reimage the instances by default when the `sku` (or select other fields, like the `source_image_reference` etc) is updated. This can be disabled by setting the `terraform_should_roll_instances_when_required` field to `false`.
* As outlined in #2807, Extensions for a Virtual Machine Scale Set will form part of a separate resource, rather than being part of the main resource. This allows operators to more easily control the lifecycle of these dependencies.

.. and a few things to note here:

* This **resource is feature-toggled** and **disabled by default** until the remaining Virtual Machine and Virtual Machine Scale Set resources become available, at which time we'll document how to enable this feature-toggle.
* This resource is in a **Beta** state - meaning that the Schema can change at any point prior to 2.0 - when it'll go GA.
* There's one [upstream issue](https://github.com/Azure/azure-rest-api-specs/pull/7355) blocking some functionality of this resource, but this is limited to Virtual Machine Scale Sets with Automatic/Rolling Upgrades configured - if this isn't resolved by the time we go GA we'll have to work around this/raise an error. In addition one test in here is also blocked on #4453.
* The examples referenced in the documentation (tracked separately [in this branch](https://github.com/terraform-providers/terraform-provider-azurerm/tree/examples/new-vmss-resources)) will be completed and PR'd once the remaining Virtual Machine/Virtual Machine Scale Set resources have been merged

---

As mentioned above - **this is feature-toggled off by default** - once the remaining Virtual Machine/Virtual Machine Scale Set resources become available we'll publicise this, but for the moment I've intentionally held off documenting how to opt into this functionality.

Upcoming PR's will add support for [the other new Virtual Machine and Virtual Machine Scale Set resources documented in the 2.0 announcement](https://github.com/terraform-providers/terraform-provider-azurerm/issues/2807).

Fixes #160
Fixes #292
Fixes #536
Fixes #809
Fixes #1616
Fixes #1620
Fixes #1669
Fixes #1739
Fixes #3020
Fixes #3773
Supersedes #2750

### CODE REVIEW

File: azurerm/internal/services/compute/virtual_machine_scale_set.go [hunk 1]
```diff
@@ -0,0 +1,1282 @@
...
+		Type:     schema.TypeList,
+		Required: true,
+		// TODO: confirm this is the same as MinItems: 1
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"name": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ForceNew:     true,
+					ValidateFunc: validation.NoZeroValues,
```
Comment (katbyte) [inline azurerm/internal/services/compute/virtual_machine_scale_set.go]: Should this be 
```suggestion
					ValidateFunc: validation.NoEmptyStrings,
```

File: azurerm/internal/services/compute/ssh_keys.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+func SSHKeysSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeSet,
+		Optional: true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"public_key": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validation.NoZeroValues,
```
Comment (katbyte) [inline azurerm/internal/services/compute/ssh_keys.go]: Should this be 
```suggestion
					ValidateFunc: validation.NoEmptyStrings,
```

File: azurerm/internal/services/compute/ssh_keys.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+				},
+
+				"username": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validation.NoZeroValues,
```
Comment (katbyte) [inline azurerm/internal/services/compute/ssh_keys.go]: Should this be 
```suggestion
					ValidateFunc: validation.NoEmptyStrings,
```

File: azurerm/resource_arm_linux_virtual_machine_scale_set.go [hunk 4]
```diff
@@ -0,0 +1,1074 @@
...
+	if err != nil {
+		return fmt.Errorf("Error retrieving Linux Virtual Machine Scale Set %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	if resp.ID == nil {
+		return fmt.Errorf("Error retrieving Linux Virtual Machine Scale Set %q (Resource Group %q): ID was nil", name, resourceGroup)
+	}
+	d.SetId(*resp.ID)
+
+	// this shouldn't need to go into Update, but let's see
```
Comment (katbyte) [inline azurerm/resource_arm_linux_virtual_machine_scale_set.go]: ?
```suggestion
	// todo this shouldn't need to go into Update, but let's see
```

File: azurerm/resource_arm_linux_virtual_machine_scale_set.go [hunk 5]
```diff
@@ -0,0 +1,1074 @@
...
+						instanceIdsToRoll = append(instanceIdsToRoll, *instance.InstanceID)
+					}
+				}
+
+				if err := instances.NextWithContext(ctx); err != nil {
+					return fmt.Errorf("Error enumerating instances: %s", err)
+				}
+			}
+
+			// there's a performance enhancement to do batches here, but this is fine for a first pass
```
Comment (katbyte) [inline azurerm/resource_arm_linux_virtual_machine_scale_set.go]:
```suggestion
			// todo there's a performance enhancement to do batches here, but this is fine for a first pass
```

File: azurerm/resource_arm_linux_virtual_machine_scale_set.go (hunk 4 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_linux_virtual_machine_scale_set.go]: this was part of scoping, turns out this can be removed

**Comment by tombuildsstuff** [comment]
Ignoring 5 test failures (which are Preview related, where we're still trying to get access, [will be fixed by this PR](https://github.com/terraform-providers/terraform-provider-azurerm/pull/4453) - or [are blocked on this issue](https://github.com/Azure/azure-rest-api-specs/pull/7355)) - the tests otherwise look good:

Screenshot 2019-10-07 at 12 32 20

Since this is feature-toggled off we're good to merge this for the moment, and can publish the information for how to enable this once all the VM/VMSS resources are available [and the upstream API issue is fixed](https://github.com/Azure/azure-rest-api-specs/pull/7355)

**Pull request merged**